src/wasm/wasm-ir-builder.cpp

```diff
@@ -634,18 +758,42 @@ Result<> IRBuilder::visitEnd() {
     iff->ifFalse = *expr;
     iff->finalize(iff->type);
     push(maybeWrapForLabel(iff));
+  } else if (auto* tryy = scope.getTry()) {
+    tryy->body = *expr;
+    tryy->finalize(tryy->type);
+    push(maybeWrapForLabel(tryy));
+  } else if (Try * tryy;
+             (tryy = scope.getCatch()) || (tryy = scope.getCatchAll())) {
+    tryy->catchBodies.push_back(*expr);
+    tryy->finalize(tryy->type);
+    push(maybeWrapForLabel(tryy));
   } else {
     WASM_UNREACHABLE("unexpected scope kind");
   }
   return Ok{};
 }
 
-Result<Index> IRBuilder::getLabelIndex(Name label) {
+Result<Index> IRBuilder::getLabelIndex(Name label, bool inDelegate) {
   auto it = labelDepths.find(label);
   if (it == labelDepths.end() || it->second.empty()) {
```

Comment:
I checked the definition of `labelDepths`, and not sure why the value needs to be not a single index but a stack of indices. I thought that is mostly for translating a name to an index?

Reply:
It's to handle shadowing of label names correctly. If we have this:

```
block $L
 block $LL
  block $L
    block $LL
      br $LL
    end
    br $LL
  end
end
```

We still need to know where the second `br $LL` should go, so we can't have completely overwritten the outer `$LL` when we reached the inner `$LL`.